### PR TITLE
Add Windows implementations.

### PR DESCRIPTION
**Pull request by kazatsuyu**
The following points should be noted:

* `cargo mobile update` has not been tested, but it will probably fail on Windows because the running binary cannot be overwritten.
* Use hard links for libraries under jniLibs since symbolic links cannot be created, and simply copy them if they fail.
* Instead of symbolic links for assets directory, copy them at Gradle build time.

### CODE REVIEW

**Comment by kazatsuyu** [comment]
Can someone please review it?

**Comment by Zeerooth** [comment]
oh god, I'm so sorry, I wanted to enable the windows CI on your branch, but I accidentally pushed some changes from my other branch, I'll try to have that sorted out

File: src/os/macos/mod.rs [hunk 1]
```diff
@@ -35,6 +35,7 @@ impl Display for DetectEditorError {
 pub enum OpenFileError {
     PathToUrlFailed { path: PathBuf },
     LaunchFailed(OSStatus),
+    BossyLaouchFailed(bossy::Error),
```
Comment (Zeerooth) [inline src/os/macos/mod.rs:38]: should be `BossyLaunchFailed` instead

File: src/os/windows/mod.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            PWSTR(ANDROID_STUDIO_UNINSTALL_KEY_PATH.as_ptr() as _),
+            PWSTR(ANDROID_STUDIO_UNINSTALLER_VALUE.as_ptr() as _),
+            PWSTR(buffer.as_mut_ptr()),
+            0,
+        )
+    };
+    if lstatus.0 as u32 != ERROR_SUCCESS.0 {
+        return Err(OpenFileError::OSError(windows::Error::from_win32()));
+    }
+    let len = NullTerminatedWTF16Iterator(buffer.as_slice().as_ptr()).count();
```
Comment (Zeerooth) [inline src/os/windows/mod.rs:182]: `as_slice()` on arrays is a nightly feature, you should probably use `&buffer[..]`

**Review by zeerooth — CHANGES_REQUESTED** [review]

File: templates/platforms/android-studio/app/build.gradle.kts.hbs [hunk 3]
```diff
@@ -61,12 +61,29 @@ dependencies {
     implementation("{{this}}"){{/each}}
 }
 
+{{#if windows}}tasks.register<Sync>("syncAssets") {
```
Comment (Zeerooth) [inline templates/platforms/android-studio/app/build.gradle.kts.hbs:64]: I see a major inconvenience with specifying `gradle` tasks like that - if you transfer this project from a windows machine to some other OS (and vice-versa) this won't get updated and the build will most probably fail. So it's a good idea overall to leave the android project agnostic of its host. But I'm not sure what the solution could be - would it be possible to use window's hard links instead of copying the files?

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs:64]: Hard links are not permitted for directory. I thought about using junctions, but when I move to another location, it can only be treated as a normal directory, which defeats the purpose of having the contents of the two directories synchronized. The creation of a symlink for the assets directory is only done at initialization time, so even if a project initialized on Windows is transferred to another OS, it will be copied as much as necessary at build time, although it is not efficient, and should not fail. On the other hand, if a project that was initialized on a non-Windows OS is brought to Windows, it will fail, but I haven't come up with an efficient solution for that. Would it be possible to remove the `{if windows}}` and use the `Sync` task instead of link on all platforms? Of course, that would add some overhead.

File: src/os/macos/mod.rs (hunk 1 above)
Comment (kazatsuyu) [reply src/os/macos/mod.rs]: Fix it.

File: src/os/windows/mod.rs (hunk 2 above)
Comment (kazatsuyu) [reply src/os/windows/mod.rs]: Oh, no. I didn't need to make it a slice, so I deleted it.

**Review by zeerooth — COMMENTED** [review]
I think that generally this PR is very sound and well-made, good job!

But, before merging there are a couple of things to sort out:
- We need to resolve the issue with symlinking the assets directory, [as discussed here](https://github.com/BrainiumLLC/cargo-mobile/pull/45#discussion_r742277129)
- One it's done I'd be nice to test the basic commands like `cargo mobile init`, `cargo mobile build`, `cargo mobile open` etc. across all supported platforms, because there are changes in this PR that indirectly affect Linux and Mac as well
- Update the README (add Windows to supported platforms and mention any Windows' specific caveats like `cargo mobile update` failing)

File: templates/platforms/android-studio/app/build.gradle.kts.hbs (hunk 3 above)
Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: The assets directory is now linked at build time instead of at the time of project generation.

1. Try symbolic linking.
2. If it fails, try hard linking all files in the assets directory.
3. If all else fails, copy them.

On maxOS and Linux, a symbolic link will be generated, on Windows, hard links will be generated unless the volume is different or some other reason.

**Comment by kazatsuyu** [comment]
I think I've solved the problem of symbolic links. 

In addition, I fixed a path problem in stable that I missed using nightly.

I've confirmed that the commands work on Windows except cargo update, but I haven't checked if there are no problems on Linux and macOS. (Of course, it would be better to have someone other than me test it on Windows as well.)

File: templates/platforms/android-studio/app/build.gradle.kts.hbs (hunk 3 above)
Comment (Zeerooth) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: I tested your changes on Linux and it seems like the symbolic link is broken 
```
ls

assets  Cargo.lock  Cargo.toml  gen  mobile.toml  README.md  src  target
```
```
cd gen/android/test-windows/app/src/main
file assets

assets: broken symbolic link to ../../../../assets
```
As if the symlink was generated a few levels too low.

Another thing is that I now know why the assets are better to be generated on project creation.
When you open the project in Android Studio with `cargo android open` you lose the access to the assets directory unless you run a build. This is quite problematic since cargo-mobile is specifically designed to work well with Android Studio. Also, if for whatever reason the symlinking of asset directories fails for the first time, the `LinkAssets` task will repeatedly create hard links (or copy the files) even if the operating system supports them. I suppose it would be wiser to just print an error when the symlinking fails rather than do these fallbacks and possibly desync the assets. Not to mention that it still doesn't solve the issue of project portability between different systems since the hard links will get broken.

I've read a bit on the subject of hard links and junctions in Windows' and I found out that actually, Windows supports soft links ! Granted, they were introduced in Vista and [used to require the admin pivileges](https://en.wikipedia.org/wiki/NTFS_links#Restrictions_and_drawbacks) but in [Windows 10 they don't anymore](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/). What do you think about using these instead? 
The drawback is that if somebody uses Windows 7, for example, and doesn't have the required privileges, it will fail. But, in that case we could just print the error and [point out the relevant information for solving it](https://stackoverflow.com/questions/23217460/how-to-create-soft-symbolic-link-using-java-nio-files#answer-24353758).
This would simplify the whole process much I think.

In concusion, the process would look like this:
1. User runs `cargo mobile init`
2. `cargo-mobile` creates the project, tries to create the symlink to `assets/` from `gen/android/test-windows/app/src/main/`
- On Linux & Mac, this works as before.
- On Windows, try to create a soft link and if the user uses a version older than Windows 10 and doesn't have the privileges, print out the error and relevant information to fix this.

Now, there's still the issue of portability - symlinks created on Windows won't be recognised on Linux and vice-versa, but I suppose that we can just write in README that you need to run `cargo mobile init` again (to regenerate the symlink) if you just transferred the project from another OS.

So that's my proposal on the direction we should take on this matter. What do you think about it @kazatsuyu?

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: I made a mistake in the origin of the relative path and created a strange symbolic link. I'll fix this soon.

Also, I hadn't noticed the problem when launching in Android Studio before creating the assets directory. How about doing the same thing at project generation and build time?

Regarding symbolic links in Windows 10, this only works if in developer mode. I tried it by passing `SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE` to [`CreateSymbolicLinkW`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinkw), but it failed in non-developer mode. I'm wondering if this will make the feature unusable for most Windows users, since not many people, even developers, are in developer mode.
If giving the privilege to create symbolic links is okay, then developer mode might be a good assumption, but it doesn't seem very attractive to me.

I need to think more about the problem of moving projects between different OS. I recognized that I need to prepare two environments to check how it works, but I'm trying it out in a Linux environment on Docker and haven't been able to build the environment properly yet. So it may take some more time.

Comment (Zeerooth) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: Thanks, the symlink indeed works now, but it wipes out all my assets. I suppose it's [because of this line?](https://github.com/BrainiumLLC/cargo-mobile/pull/45/files#diff-017aabcce71c07cbaa703c555101b6dd3fee0665a772b7592c32e2e2c1e21c96R20)

We could do the symlinking on project generation and build time and that would actually solve the issue with unaccessible assets directory in Android Studio from the start. But still, the issue on Windows would be that hard links work only one way, that is if we add a file in `assets/` it will get linked to `gen/android/test-windows/app/src/main/assets/`, sure, but not the other way as far as my understanding goes, which is the issue when you want to edit the assets in Android Studio.

I suppose you're right that having the requirement of enabling the developer mode may not be ideal for some users on Windows, but we might need to do this if we want to limit the complexity of our soft link/hard link/copy logic.

I looked around some other projects that support Windows to see how they handle symlinks and... I haven't found much, only that `fnm` [used to require developer mode to work on Windows](https://github.com/Schniz/fnm/commit/6e9f085c656b7e613e31607431293da729098059#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) but [symlinks were replaced with junctions](https://github.com/Schniz/fnm/pull/519). Unfortunately, this probably doesn't work for us since, as you've mentioned before, junctions use absolute paths and won't work when moving the project.

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: I've been making some modifications to allow symlinks to work on Windows, and two things have been bothering me.
1. The `-F` option of the `ln` command does not specify overwriting a directory, but I think [it's used for that purpose](https://github.com/BrainiumLLC/cargo-mobile/blob/master/src/util/ln.rs#L142). 
   ```
   -d, -F, --directory         allow the superuser to attempt to hard link
                               directories (note: will probably fail due to
                               system restrictions, even for the superuser)
   ```
   If `Clobber::FileOrDirectory` means what I take it to mean, shouldn't `std::fs::remove_dir_all` be called?
2. I thought the `assets` directory link was created in `gen/android/<package-name>/app/src/main/assets`, but it is actually created in `gen/android/<package-name>/app/src/main/assets/assets`. Is this the intended behavior?

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs:64]: I've decided not to consider `Clobber::FileOrDirectory` since there doesn't seem to be a case where it is actually used. It is likely that symbolic links will become directories when projects are exchanged between different OSes, so a mechanism to forcibly overwrite directories seems like a good idea, but it is not included in the fix for now.

The issue of links being created in `assets/assets` is also left unchanged. This doesn't seem to be a matter for Windows, even if it should be fixed.

If symbolic links are available, there is no need to do hard linking at build time, so I changed that too.

Most of the parts related to linking on Linux and macOS are now back to the original version.

Comment (Zeerooth) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: Thanks for making these changes!

When it comes to the `-F`/`-f` switch for the `ln` command I think there might be disparity between the Mac and Linux, because on Mac, the `-F` flag does something entirely different:
```
-F    If the target file already exists and is a directory, then remove it so that the link may occur.  The -F option should be used with
      either -f or -i options.  If none is specified, -f is implied.  The -F option is a no-op unless -s option is specified.
```
You can read the whole man page for `ln` on Mac online [here](https://www.unix.com/man-page/mojave/1/ln/)
I think that the enum option of `Clobber::FileOrDirectory` needs to be either removed entirely because it's not being used currently (or more preferably) make the flags conditional based on the compilation target. Though, you don't have to fix that in this PR because I guess it is kind of out of scope...

As for the `assets/assets` issue, I'm pretty much sure this is a bug because in the current state the `bevy` template can't load the default `branding/icon.png` asset and throws the following error error:
```
5133  5184 I RustStdoutStderr: thread 'IO Task Pool (1)' panicked at 'called `Result::unwrap()` on an `Err` value: PathLoaderError(NotFound("branding/icon.png"))', /home/radek/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_asset-0.3.0/src/asset_server.rs:295:60
```
If I remove one level of redundant `assets` it works fine, so if you can please add a fix for that.

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: I see. The `ln` command seems to be different between Linux and BSD.

I fixed the problem of the `assets/assets` directory, but now `cargo mobile init` fails for projects generated with earlier versions. The directories that were already generated need to be removed. Should I make `Clobber::FileOrDirectory` available?

Comment (Zeerooth) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: Yes, in that case `Clobber::FileOrDirectory` is necessary I guess. However, the problem is that this still won't work for Linux because `ln` can only force removing files by adding `-f`, doesn't have a way to remove directories and errors out with `cannot overwrite directory`... So I guess the only option would be to first remove the target directory and then perform the linking.

Comment (kazatsuyu) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: Directories that already exist will now be removed. Also, use `remove_dir_all` instead of `-F` option even on macOS to avoid OS-specific branching.

Comment (Zeerooth) [reply templates/platforms/android-studio/app/build.gradle.kts.hbs]: Thanks, looks good now, I think we can resolve that thread :D

**Review by zeerooth — APPROVED** [review]

**Review by zeerooth — CHANGES_REQUESTED** [review]

File: src/env.rs [hunk 4]
```diff
@@ -22,23 +22,42 @@ impl Reportable for Error {
...
+    tmp: Option<String>,
+    temp: Option<String>,
+    program_data: Option<String>,
 }
 
 impl Env {
     pub fn new() -> Result<Self, Error> {
-        let home = std::env::var("HOME").map_err(Error::HomeNotSet)?;
+        let home = std::env::var("HOME");
+        let userprofile = std::env::var("USERPROFILE");
```
[the remark this answers is not in the document]
Comment (kazatsuyu) [reply src/env.rs]: I added `os::windows::env::Env` struct and reset `env::Env`.

**Review by zeerooth — DISMISSED** [review]

**Comment by Zeerooth** [comment]
Once again, thank you so much for your contribution!
I think that now this PR is ready to be merged - I've briefly tested the latest version on Linux and Mac, updated the README and approved it, but since the changes here are quite massive, I'd like someone else to take a look and approve as well.

@francesca64 may I kindly ask you to take a look?

**Review by zeerooth — COMMENTED** [review]
I wanted to merge this but I encountered a weird behaviour when testing on this branch now...
When I run `cargo mobile init` for the first time the project is created successfully. However, if I do it the second time, I get the following error:
```
ln: failed to create symbolic link '/home/radek/code/cargo-mobile-sandbox/a/gen/android/a/app/src/main/assets': No such file or directory
error: Asset dir couldn't be symlinked into Android project
    Failed to create a symbolic link from "../../../../../../assets" to directory "/home/radek/code/cargo-mobile-sandbox/a/gen/android/a/app/src/main/" (file and directory clobbering enabled): `ln` command failed: Command
    "ln -n -s -f ../../../../../../assets /home/radek/code/cargo-mobile-sandbox/a/gen/android/a/app/src/main/assets" didn't complete successfully, exiting with code 1.
```
and the `gen/android/a/app/src/main` directory appears to be missing.
When I then run `cargo mobile init` again, the project is initialized successfully.
Basically, this behaviour alternates and it needs to be fixed.

**Review by francesca64 — CHANGES_REQUESTED** [review]
Thanks for making such a massive contribution!

(and thanks to @Zeerooth for thoroughly reviewing it!)

**Comment by amrbashir** [comment]
If there is any work needed for this PR to be merged, I'd be glad to help (ofc, If the author of this PR gives me permissions to, or if he is inactive, I could create a new PR that is build on top of this PR and keep his commits in the new PR)
